# Greatest Common Divisor

The Greatest Common Divisor (GCD), sometimes known as the highest common factor, is the largest number which divides two positive integers (a,b).

For `a = 12, b = 8` we can calculate the divisors of a: `{1,2,3,4,6,12}` and the divisors of b: `{1,2,4,8}`. Comparing these two, we see that `gcd(a,b) = 4`.

Now imagine we take `a = 11, b = 17`. Both `a` and `b` are prime numbers. As a prime number has only itself and `1` as divisors, `gcd(a,b) = 1`.

We say that for any two integers `a,b`, if `gcd(a,b) = 1` then `a` and `b` are coprime integers.

If `a` and `b` are prime, they are also coprime. If `a` is prime and `b < a` then `a` and `b` are coprime.

> Think about the case for `a` prime and `b > a`, why are these not necessarily coprime?


There are many tools to calculate the GCD of two integers, but for this task we recommend looking up [Euclid's Algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm).

Try coding it up; it's only a couple of lines. Use `a = 12, b = 8` to test it.

Now calculate `gcd(a,b)` for `a = 66528, b = 52920` and enter it below.

## Solution

In [1]:
from math import gcd

a = 66528
b = 52920

print(gcd(a, b))

1512


### Flag: 1512

# Extended GCD

Let a and b be positive integers.

The extended Euclidean algorithm is an efficient way to find integers u,v such that

a * u + b * v = gcd(a,b)

> Later, when we learn to decrypt RSA, we will need this algorithm to calculate the modular inverse of the public exponent.


Using the two primes p = 26513, q = 32321, find the integers u,v such that

p * u + q * v = gcd(p,q)

Enter whichever of u and v is the lower number as the flag.

> Knowing that p,q are prime, what would you expect gcd(p,q) to be? For more details on the extended Euclidean algorithm, check out [this page](http://www-math.ucdenver.edu/~wcherowi/courses/m5410/exeucalg.html).

## Solution

In [2]:
from sage.all import *

p = 26513
q = 32321
print(xgcd(p, q))

(1, 10245, -8404)


### Flag: -8404

# Modular Arithmetic 1

<p align="center">
    <img src="./images/modular1.png">
</p>

## Solution

In [3]:
x = 11 % 6
y = 8146798528947 % 17
print(min(x, y))

4


### Flag: 4

# Modular Arithmetic 2

<p align="center">
    <img src="./images/modular2.png">
</p>

## Solution

In [4]:
p = 65537
e = 65536
a = 273246787654

assert e == p - 1
# => fermat

print(pow(a, e, p))

1


### Flag: 1

# Modular Inverting

<p align="center">
    <img src="./images/invert.png">
</p>

## Solution

In [5]:
print(pow(3, -1, 13))

9


### Flag: 9